### ISSUE TITLE
dict.pop() causes no-matching-overload

### ISSUE THREAD

**Issue by noamraph** [astral-sh/ty#3819]
### Summary

This code fails type checking, and I expect it to pass:

```python
def validate_type(data: object) -> object:
    if isinstance(data, dict):
        type_ = data.pop('type', None)
```

Shows:
```
No overload of bound method `dict.pop` matches arguments (no-matching-overload) [Ln 6, Col 17]
info: First overload defined here: First overload defined here [Ln 3155, Col 5]
info: Possible overloads for bound method `pop`:
info:   (self, key: _KT@dict, /) -> _VT@dict
info:   (self, key: _KT@dict, default: _VT@dict, /) -> _VT@dict
info:   [_T](self, key: _KT@dict, default: _T, /) -> _VT@dict | _T
```

See: https://play.ty.dev/1190e0ae-800d-4425-b20e-650a12b13126

I just noticed that this fails also without the default value, with another issue:

```python
def validate_type(data: object) -> object:
    if isinstance(data, dict):
        type_ = data.pop('type')
```

Shows:

```
Argument to bound method `dict.pop` is incorrect: Expected `Never`, found `Literal["type"]` (invalid-argument-type) [Ln 3, Col 26]
info: Matching overload defined here [Ln 3156, Col 9]
Parameter declared here [Ln 3156, Col 19]
info: Non-matching overloads for bound method `pop`:
info:   (self, key: _KT@dict, default: _VT@dict, /) -> _VT@dict
info:   [_T](self, key: _KT@dict, default: _T, /) -> _VT@dict | _T
```

See https://play.ty.dev/bf9f06d4-631a-4af8-b70d-76eba1752fcb

### Version

ty 0.0.51

**Comment by sharkdp** [comment]
After `isinstance(data, dict)`, the type of `data` is narrowed to `object & Top[dict[Unknown, Unknown]]` which simplifies to `Top[dict[Unknown, Unknown]]`. This type represents "some dictionary", but we know nothing about it's key and value type, and so when you try to use `.pop` with a string key, that operation could fail (it could be a dictionary with keys of another type). You can use `data: Any` or a custom TypedDict type to avoid that. You can also use an operation that wouldn't lead to a `TypeError`, like `data.get`

**Comment by carljm** [comment]
IMO it should not be a type error to use `.pop()` with a default value, regardless of key type. So just like [we patch typeshed](https://github.com/astral-sh/ruff/pull/25831) so that the first argument of `dict.get` is typed as `object` rather than `KT`, I believe the [overloads of `dict.pop`](https://github.com/python/typeshed/blob/c9e422dbac410054732ae2006fa088db44f7d07d/stdlib/builtins.pyi#L1339-L1344) that include a default should have the key typed as `object`, not `KT`.

**Comment by noamraph** [comment]
@carljm thanks for this lightning speed fix! Amazing!

I thought about it, and I think that there may be a more precise definition of when pop() should fail static typing. I don't know if ty's type system supports it.

First, a complication: I actually think that pop() should not really deserve a special case for when a default is given. I know that catching KeyErrors is pretty idiomatic in Python. So, it would have been pretty idiomatic to write my existing code:

```python
    def validate(cls, data: object) -> object:
        if isinstance(data, dict):
            type_ = data.pop('type', None)
            if type_ and type_ != 'STEP':
                raise ValueError(f'Expecting type to be STEP, is {type_!r}')
        return data
```

like this:

```python
    def validate(cls, data: object) -> object:
        if isinstance(data, dict):
            try:
                type_ = data.pop('type')
            except KeyError:
                pass
            else:
                if type_ != 'STEP':
                    raise ValueError(f'Expecting type to be STEP, is {type_!r}')
        return data
```

Which still fails type checking after the change.

I think that `pop()`, and even `__getitem__`, should fail type checking when we *know* that the key couldn't be in the dict. I think this means: there is no overlap between the key type and the type of the argument.

So this should fail (after your fix it doesn't fail):
```python
def pop(d: dict[int, object]):
    d.pop("hello", None)
```

But this should not fail (currently it does):
```python
def pop(d: dict[int | bool, object], x: int | str):
    d.pop(x)
```

WDYT?

Thanks!
Noam

**Comment by carljm** [comment]
@noamraph That's entirely reasonable, and I agree it would be the ideal behavior. But it requires a type system feature that doesn't currently exist (an annotation like `Overlaps[KT]` or similar) in order to annotate `dict.get` and `dict.pop` in this way. There has been some discussion of adding this feature, and in future we may experiment with it in ty. For now I think the (arguable, since the code really doesn't error at runtime) false negative on non-overlapping types is preferable to false positives.

Regarding the no-default-provided case, this is a much more general issue in Python typing. Type checkers will typically error if you do something non-type-safe that may raise an exception. It is always possible that this code is wrapped in a try block which will catch and handle that exception; type checkers do not attempt to detect this (and it would be quite hard to do so in a general way). If you write code in this style, the best option IMO is to explicitly suppress the type error, noting that the exception will be caught and handled.

**Comment by noamraph** [comment]
Thanks @carljm , it's good to know that what I said was reasonable!

I agree that as long as there's no `Overlaps[KT]` feature, the behavior after your change seems the best.

Regarding the no-default-provided case: I have the feeling that KeyError is different from other exceptions. I think it is more like another option for a return value - something like `Option<T>`, that Python lacks. The official Python glossary even has an [entry about it](https://docs.python.org/3/glossary.html#term-EAFP). It's also more efficient than doing two lookups, first for checking if the key exists and then for getting the value. This is specific to KeyError - I don't think the type checker should handle other exceptions like this. So I'm just saying that if the `Overlaps` feature is added, I think it should be used also for the no-default case.

Thanks again!
Noam

**Comment by carljm** [comment]
@noamraph Are you suggesting that you would want the typechecker to allow `dict[str, int].pop(typed_as_object)`, regardless of whether that line is actually wrapped in a `try/except KeyError:`? Or only if that line is in fact wrapped in a `try/except KeyError:`?

**Comment by noamraph** [comment]
@carljm no, I don't think that the typechecker should look for try..except. I just think that the key in no-default calls should be treated the same way as with-default calls, once you have `Overlaps[KT]`.

**Comment by carljm** [comment]
Hmm, I don't think I agree with that.

**Comment by noamraph** [comment]
It's just theoretic at this stage, but I think that actually, always using `Overlaps[KT]`, even when there's no default, is fine. First, I think that as trying to deal with `try..except KeyError` is different from anything else that ty does, then it's probably not a good idea to try. Anyway I guess it will only be heuristics - for example, it's perfectly fine to define a function, like I did in the example, where the intended behavior is to propagate the KeyError.

But mostly, I think that this would give reasonable results. If KT is `object`, then obviously it would be fine. I think that most other types won't overlap unless one is a subclass of the other. Say KT is `str`. Then for most types which are not `str` subtypes, type checking would fail. For `object` it would pass, and I actually think it's reasonable: the object could be a string which could be in the dict, it could be a string which is not in the dict and you'll get a KeyError, and it could also be something else which would not be in the dict and you'll get a KeyError.

If the guiding principle is to try to avoid false positives, then I think that using `Overlaps[KT]`, which fails exactly in the cases where it's certain that there is a bug, is the right thing to do.

**Comment by carljm** [comment]
I guess the critical distinction is that `KeyError` is always possible, even if you have exactly the right type, since a `dict` type doesn't model presence or absence of specific keys. So if it's OK for `KeyError` to be possible (but not guaranteed) when you `dict[str, int].pop(typed_as_str)`, it's arguably just as fine for `KeyError` to be possible (but not guaranteed) when you `dict[str, int].pop(typed_as_object)`.

I do find that pretty convincing, thanks! Will consider this when we try the `Overlaps` route.
